### ISSUE TITLE
[MASTER] Causal Uplift — Sprint 2+ Execution Plan

### ISSUE THREAD

**Issue by arthur105204** [arthur105204/causal-uplift-modeling#20]
# Goal

Authoritative execution overview for `causal-uplift-modeling`. Every other Issue is a small,
self-contained execution spec that inherits the rules on this page. Read this Issue once; then
work Issue-by-Issue without re-reading governance documents.

Architecture: **Kaggle-first + notebook-first + feasibility-first + learning-first.**

# Execution architecture

```text
Kaggle
├── primary notebook execution
├── saved versions
├── full-data experiments
└── presentation

GitHub repository
├── notebook source / version history
├── reusable Python source
├── configs
├── focused tests
└── documentation

src/       reusable logic only when reuse / correctness / a contract justifies extraction
tests/     focused correctness and regression tests
outputs/   predictions, metrics, manifests, and downstream inputs as justified
```

Kaggle is the primary analytical execution environment, notebook runtime, and presentation
surface. GitHub remains task management, version control, and the architecture/specification
record.

**Notebook-first is not notebook-only, and it is not notebook-hidden.** The scientific
narrative — question, method, result, interpretation, downstream implication — stays visible in
the notebook. Engineering plumbing (hashing, manifest writing, environment capture,
serialization) belongs in `src/` or an appendix. Do not hide research logic behind opaque
helper APIs, and do not create abstractions for architectural appearance.

# Task vs notebook

An Issue is an **execution task**. A notebook is an **artifact**. They are not 1:1. Several
Issues write into the same notebook; each stays a separate task with its own checklist and
Definition of Done.

| Notebook | Issues |
|---|---|
| `00_project_overview.ipynb` | MASTER #20 — project narrative |
| `01_data_feasibility.ipynb` | S01 #1, T01 #2, T03-A #4 (essential pre-model integrity) |
| `02_eda_split_preprocessing.ipynb` | T02 #3, T05 #6, T04 #5, T03-C #4 (deferred calibration) |
| `03_metrics_and_meta_learners.ipynb` | T06 #7, T07 #8, T08 #9, T09 #10 |
| `04_causal_forest.ipynb` | T10 #11, T11 #12 |
| `05_validation_uncertainty.ipynb` | T12 #13, T13 #14 (conditional), T14 #15 (P1), T15 #16 |
| `06_heldout_evaluation.ipynb` | T16 #17, T17 #18 |
| `07_final_story.ipynb` | T18 #19 |

The legacy notebooks (`01_sprint1_feasibility_evidence`, `02_data_engineering_pipeline`,
`03_exploratory_data_analysis`, `04_data_quality_causal_integrity`) are **historical evidence**
from the pre-reset architecture. They are retained, not deleted. Their accepted results stay
valid and are inherited rather than re-derived. The V3 numbering above is by notebook role, not
by task number.

# Critical path (P0)

```text
architecture / data feasibility
→ EDA
→ frozen split
→ minimal preprocessing
→ uplift metrics
→ Random + Response baseline
→ T-Learner
→ X-Learner
→ Causal Forest
→ common validation comparison
→ bootstrap uncertainty
→ pre-test freeze
→ one-shot held-out evaluation
→ final interpretation
```

```text
#1 S01 → #2 T01 → #4 T03-A → #3 T02 → #6 T05 → #5 T04 → #7 T06 → #8 T07
→ #9 T08 → #10 T09 → #11 T10 → #12 T11 → #13 T12 → #16 T15 → #17 T16
→ #18 T17 → #19 T18

off the blocking path, in parallel after #6 T05:
#4 T03-C  (mandatory; must complete before #17 T16)
#15 T14   (P1; after #13 T12)
#14 T13   (conditional; only if #13 T12 is surprising or unstable)
```

Nothing outside this path may block it.

## P0 — mandatory, blocking

S01, T01, T03-A, T02, T05, T04, T06, T07, T08, T09, T10, T11, T12, T15, T16, T17, T18.

## P0 — mandatory, non-blocking

**T03-C** (2,000-draw randomization calibration). Mandatory and frozen. It is *not* a
predecessor of T04 or of any modeling task, because `docs/06` Stage 0 item 3 states the
calibration algorithm is frozen early while "generated development-data thresholds are locked
later, before the pre-test executable freeze," and `docs/03` requires the generated percentiles
to be finalized before the pre-test executable freeze. Run it in parallel once T05 exists.
Deadline: before T16.

## P1 / conditional / deferred

| Item | Class | Rule |
|---|---|---|
| T13 root-cause diagnostics | Conditional | Execute only when T12 produces a surprising or unstable result. Not an unconditional predecessor of T15. |
| T14 decile/segment analysis | P1 | After T12; runs in parallel with downstream P0 work. Never blocks T15/T16/T17. |
| DR-Learner | Stretch / conditional | Development-only. Enters held-out evaluation only if every promotion gate passes before freeze. |
| S-Learner | Deferred | Not implemented; never silently substituted for Response or T-Learner. |
| Cross-language CF bridge | Conditional | Python-only unless documented evidence proves a required capability is missing (D22). |
| Extensive hyperparameter tuning | Noncritical | Bounded predeclared candidates only. |
| Decorative EDA | Noncritical | Every retained output must answer a stated question. |
| Excessive infrastructure / governance | Noncritical | Must not delay the mandatory path. |

**Causal Forest is a mandatory accepted main comparator (D15).** It may not be omitted,
substituted, or deferred for difficulty, runtime, or unfavorable performance. If every eligible
exact implementation fails its required gates, retain the failure evidence and STOP for explicit
owner methodology review.

# Incremental execution model

Each Issue carries small numbered sub-tasks. The intended loop is:

```text
read Issue → implement one or a few unchecked sub-tasks → verify → tick the checklist → next
```

Do not implement an entire task in one uncontrolled change. Do not tick a box that was not
verified.

# Frozen rules (inherited by every Issue)

These are not repeated in individual Issues. They always apply.

- `X` is exactly `f0`–`f11`, in that order, at primary `float64` precision (D09). Nothing else
  ever enters `X`.
- `T = treatment` (assignment / ITT, not exposure). Primary `Y = conversion`; `visit` is a
  secondary outcome in a separate pipeline.
- `exposure` is audit-only: excluded from `X`, from eligibility, from filtering, and from the
  treatment definition.
- `_source_row_id` is a zero-based CSV row ordinal used for provenance and alignment. It is not
  a person identifier and never enters `X`.
- Estimand is assignment/ITT CATE for ranking over the frozen eligible released-row population.
  **Predicted uplift is not a true individual treatment effect.** Empirical PEHE against true
  ITE is unavailable on this dataset and must not be reported (D28).
- Keep every released row by default (D07). Dedup/weighting is sensitivity-only; rows may be
  excluded only on a predeclared hard-integrity gate failure.
- Split: 70/15/15, joint `(T,Y)` stratification, seed `42`, source-row identity disjoint.
- **Held-out data is untouchable until T17.** No unofficial held-out diagnostics, no tuning on
  test-derived information. If an operation would read held-out performance, stop.
- Frozen portfolio: theoretical/seeded Random reference, Response LightGBM diagnostic baseline,
  T-Learner, X-Learner, mandatory Causal Forest.
- Primary ranking statistic is Qini above theoretical random (D24). Raw Qini/AUUC and
  top-K/incremental conversions are secondary (D25/D26). Response AUC/AP/log loss are
  diagnostic only and never select a causal winner (D27).
- Uncertainty: exactly 500 paired treatment-arm-stratified bootstrap draws with fixed
  predictions (D29). T17 is the first and only held-out evaluation, under the T16 freeze.
- D23 scale rungs `50K → 500K → 2M → full` still apply. Execute only the current rung; promote
  only after its predeclared correctness/resource evidence passes.
- Balance diagnostics do not prove randomization. `f0`–`f11` are anonymized and carry no known
  business meaning — do not invent one.

# Data feasibility standards

- Kaggle is the primary benchmark environment. Do not encode a specific local machine's limits
  as a universal feasibility constraint.
- Raw `.csv.gz` is the provenance/ingestion source and the checksum anchor. It is not the
  repeated analytical storage format.
- Parquet (or another typed columnar representation) is the analytical store (D19).
- Avoid repeated full-data scans. Prefer one vectorized multi-column aggregation pass over many
  single-column passes.
- Cache and reuse expensive intermediate results; persist them as run artifacts when they feed
  a downstream task.
- Sampling is acceptable for **secondary** diagnostics when statistically justified and clearly
  labeled. Per `docs/03`, a row-limited check must be labeled as such and cannot support a
  full-data conclusion — so a sampled result may not be recorded as the evidence for a
  hard gate or for a required audit row.
- Optimize for scalable reasoning, not brute-force execution.

# Source code standards

```text
notebook   visible scientific narrative
src/       reusable, testable logic
tests/     targeted correctness verification
configs/   consequential reproducibility settings
```

Extract into `src/` when logic is reused across notebooks/tasks, is mathematically or
statistically important enough to deserve a test, or would otherwise be duplicated. Do not
extract for appearance.

# Evidence before optimization

Measure before you optimize; verify before you tune; report before you interpret. A run that
executed without error is not a verified run. Notebook exists ≠ notebook parses ≠ tests pass ≠
clean Run All passes. Where applicable, acceptance requires a fresh kernel, Run All in order,
zero uncaught errors, expected outputs, and invariant reconciliation. Bounded verification may
replace an expensive rerun only when the reason and evidence are stated explicitly.

A disappointing result is not a software defect. Only a demonstrated defect justifies a rerun,
and failed-run evidence is retained, never deleted.

# Run-artifact governance

Machine-readable evidence lives in immutable `outputs/runs/<run_id>/{audit,tables,figures,
predictions,metrics,models}/`. A rerun mints a new run id; it never mutates an existing run.
Stable copies outside `outputs/runs/<run_id>/` are inspection conveniences only and never
replace historical run evidence.

# Data path

```text
configs/data_manifest.json (environment-local selector)
  → validate source identity (raw SHA-256, exact header, sizes)
  → convert to Parquet, appending _source_row_id
  → validate processed Parquet (schema, layout, ordinal continuity, semantics)
  → open processed dataset (fails closed unless the expected SHA-256 matches)
```

Never bypass a link; never discover input by glob, filename order, or directory heuristic. The
manifest is environment-local — its **paths** differ between Kaggle and a local checkout, but
its **checksums are environment-invariant**, so identity checking holds in both. Semantic
identity is mandatory; byte-identical physical output is claimed only within one environment.

# Source precedence

1. `docs/decision_register.csv` (D01–D29, owner-approved)
2. Numbered contracts `docs/01`–`docs/07` and ACCEPTED ADRs
3. Accepted empirical evidence
4. GitHub Issue wording
5. Existing code

Issues are execution views. They are reconciled to higher-authority sources, never the reverse.
If an Issue conflicts with a frozen decision, the decision wins and the Issue is corrected.

# Working phases

`[TUTOR]` → `[CODE PLAN]` → `[IMPLEMENT]` → `[VERIFY]` → `[CLEAN RUN]` → `[REVIEW]` →
`[ACCEPT]`. Execute only the requested phase; do not auto-advance. `[REVIEW]` reports findings
and does not edit unless asked. Acceptance is separate from implementation; commit, push, Issue
closure, and starting the next task are separate explicit actions.

Before non-trivial implementation the owner should be able to state INPUT, OUTPUT, the main
data/model flow, its 5–10 major steps, and the important leakage/statistical failure modes. The
owner is not expected to reinvent generic engineering infrastructure.

# Board status

**Sprint 2 — data and engineering**

- [x] [S01 — Sprint-1 feasibility evidence](https://github.com/arthur105204/causal-uplift-modeling/issues/1)
- [x] [T01 — Data engineering pipeline](https://github.com/arthur105204/causal-uplift-modeling/issues/2)
- [x] [T02 — Exploratory data analysis](https://github.com/arthur105204/causal-uplift-modeling/issues/3)
- [ ] [T03 — Data quality & causal integrity](https://github.com/arthur105204/causal-uplift-modeling/issues/4) — A: P0 blocking · C: P0 non-blocking
- [ ] [T05 — Frozen split](https://github.com/arthur105204/causal-uplift-modeling/issues/6)
- [ ] [T04 — Preprocessing strategy](https://github.com/arthur105204/causal-uplift-modeling/issues/5)
- [ ] [T06 — Uplift metrics](https://github.com/arthur105204/causal-uplift-modeling/issues/7)

**Sprint 3 — modeling**

- [ ] [T07 — Random & Response baselines](https://github.com/arthur105204/causal-uplift-modeling/issues/8)
- [ ] [T08 — T-Learner](https://github.com/arthur105204/causal-uplift-modeling/issues/9)
- [ ] [T09 — X-Learner](https://github.com/arthur105204/causal-uplift-modeling/issues/10)
- [ ] [T10 — Causal Forest verification](https://github.com/arthur105204/causal-uplift-modeling/issues/11)
- [ ] [T11 — Causal Forest training](https://github.com/arthur105204/causal-uplift-modeling/issues/12)

**Sprint 4 — validation and uncertainty**

- [ ] [T12 — Common validation comparison](https://github.com/arthur105204/causal-uplift-modeling/issues/13)
- [ ] [T13 — Root-cause diagnostics](https://github.com/arthur105204/causal-uplift-modeling/issues/14) — conditional
- [ ] [T14 — Decile & segment analysis](https://github.com/arthur105204/causal-uplift-modeling/issues/15) — P1
- [ ] [T15 — Bootstrap uncertainty](https://github.com/arthur105204/causal-uplift-modeling/issues/16)

**Sprint 5 — final evaluation**

- [ ] [T16 — Pre-test freeze](https://github.com/arthur105204/causal-uplift-modeling/issues/17)
- [ ] [T17 — One-shot held-out evaluation](https://github.com/arthur105204/causal-uplift-modeling/issues/18)
- [ ] [T18 — Final interpretation & defense](https://github.com/arthur105204/causal-uplift-modeling/issues/19)

# Open owner decisions

These are flagged, not decided. The architecture reset did **not** change any of them.

1. **T03 sampled distributional diagnostics.** `docs/03` ED-03 requires per-feature quantiles,
   quantile differences, and the empirical CDF/KS distance. `docs/03` also states a row-limited
   check cannot support a full-data conclusion. Recording a *sampled* ED-03 distributional
   result as the audit evidence therefore needs owner approval and a `docs/03` amendment. The
   current Issue wording keeps these full-data but requires a single vectorized pass.
2. **2,000-draw calibration count.** Frozen by `docs/03` (initial 2,000 draws from master seed
   42) and `docs/06` (diagnostic null calibration row). Reducing the initial count would need
   owner approval and a document amendment. It has **not** been reduced — only rescheduled off
   the blocking path, which `docs/06` Stage 0 already permits.
3. **Residual sequencing risk from that rescheduling.** `docs/03` audit ordering item 6 asks for
   material diagnostics to be bounded before the affected estimator is promoted. If T03-C runs
   very late and returns `MATERIAL_CONCERN`, the predeclared `ROBUSTNESS_REQUIRED` sensitivity
   would land after promotion. Mitigation adopted here: schedule T03-C immediately after T05, in
   parallel with T06–T09, rather than at T16.
4. **Legacy notebook placement.** Whether to move the four pre-reset notebooks into
   `notebooks/legacy/` or leave them in place under their current names. No implementation
   action has been taken.
